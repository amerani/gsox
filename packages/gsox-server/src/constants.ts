export const NOTIFICATION_TOPIC = "#";
