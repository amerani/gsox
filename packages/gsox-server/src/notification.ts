// @message('notifycommand')
// class notification {

//       @str()
//       id: string;

//       @str()
//       type: string;

//       @str()
//       timestamp: string;
// }
