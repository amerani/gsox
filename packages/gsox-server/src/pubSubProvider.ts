import { PubSub, PubSubEngine } from "graphql-subscriptions";

export const pubSub: PubSubEngine = new PubSub();
