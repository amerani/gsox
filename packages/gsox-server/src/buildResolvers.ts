import "reflect-metadata";
import { withFilter } from "graphql-subscriptions";
import { NOTIFICATION_TOPIC } from "./constants";
import { pubSub } from "./pubSubProvider";

function buildResolvers() {
      return {
            Subscription: {
                ["Notification"]: {
                      subscribe: withFilter(
                            () => pubSub.asyncIterator(NOTIFICATION_TOPIC),
                            (payload) => {
                                  if (!payload) { return false; }
                                  const notification = payload.Notification;
                                  // return notification.id === id;
                                  return true;
                            },
                      ),
                },
                Ping: {
                      subscribe: () => {
                            setInterval(() => pubSub.publish("test", {
                                Ping: { id: 0 },
                            }), 2000);
                            return pubSub.asyncIterator("test");
                      },
                },
            },
          };
}

export { buildResolvers }