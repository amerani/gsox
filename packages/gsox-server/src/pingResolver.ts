import { pubSub } from "./pubSubProvider";
import { IResolvers, IResolverObject } from "../node_modules/graphql-tools";

const pingSubscribe: IResolverObject = {
      subscribe() {
            setInterval(() => pubSub.publish("test", {
                Ping: { id: 0 },
            }), 2000);
            return pubSub.asyncIterator("test");
      },
};

const pingResolver: IResolvers<any, any> = {
      Ping: pingSubscribe
}

export default pingResolver;