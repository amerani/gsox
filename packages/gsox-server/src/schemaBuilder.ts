import { Field, Type, typeDef } from "@gsox/schema";
import { gql } from "apollo-server-express";
import { withFilter } from "graphql-subscriptions";
import { makeExecutableSchema } from "graphql-tools";
import "reflect-metadata";
import { NOTIFICATION_TOPIC } from "./constants";
import { pubSub } from "./pubSubProvider";

@Type()
class Ping {

      @Field()
      public id: string;
}

@Type()
class Notification {
      @Field(null, "Int")
      public id: number;
      @Field()
      public type: string;
      @Field("timestamp")
      public data: string;
}

const ping = typeDef(new Ping());
const not = typeDef(new Notification());

const typeDefs = gql`
      type Subscription {
            notification: Notification
            ping: Ping
      }

      ${ping}

      ${not}

      type Query {
            hello: String
      }

      type schema {
            subscription: Subscription
            query: Query
      }
`;

const id = 99;

const resolvers = {
  Subscription: {
      ["notification"]: {
            subscribe: withFilter(
                  () => pubSub.asyncIterator(NOTIFICATION_TOPIC),
                  (payload) => {
                        if (!payload) { return false; }
                        const notification = payload.notification;
                        return notification.id === id;
                  },
            ),
      },
      ping: {
            subscribe: () => {
                  setInterval(() => pubSub.publish("test", {
                      ping: { id: 0 },
                  }), 2000);
                  return pubSub.asyncIterator("test");
            },
      },
  },
};

export const schema = makeExecutableSchema({typeDefs, resolvers});
