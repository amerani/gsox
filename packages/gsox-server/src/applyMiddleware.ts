import { defaults } from "@gsox/core";
import { ApolloServer } from "apollo-server-express";
import * as bodyParser from "body-parser";
import { execute, subscribe } from "graphql";
import * as http from "http";
import "reflect-metadata";
import { SubscriptionServer } from "subscriptions-transport-ws";
import { pubSub } from "./pubSubProvider";
import { buildSchema } from "./schemaBuilder";

export function applyMiddleware(app, options) {
      const curOpt = {...defaults, ...options};
      const { host, port, routes, inject } = curOpt;

      // build schema from types
      const schema = buildSchema(inject);

      // add webhook endpoint
      app.post(routes.webhook, bodyParser.json(), (req, res) => {
            const { typeName } = req.body;
            const topic = `TOPIC_${typeName}`;
            console.log(`${topic}: ${JSON.stringify(req.body)}`);
            pubSub.publish(topic, req.body);
            return res.sendStatus(200);
      });

      // init apollo server
      const path = routes.graphql;
      const apollo = new ApolloServer({ schema });
      apollo.applyMiddleware({ app, path });

      // init subscription websocket
      const server = http.createServer(app);
      SubscriptionServer.create({
            execute, schema, subscribe,
      }, {  path, server });

      return {
            apollo,
            server,
            host,
            port,
            routes,
            listen: (listener) => server.listen({ host, port }, listener),
            close: () => server.close(),
      };
}
