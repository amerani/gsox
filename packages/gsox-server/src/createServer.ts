import { ApolloServer } from "apollo-server-express";
import * as bodyParser from "body-parser";
import { execute, subscribe } from "graphql";
import { createServer as http } from "http";
import "reflect-metadata";
import { SubscriptionServer } from "subscriptions-transport-ws";
import { NOTIFICATION_TOPIC } from "./constants";
import { pubSub } from "./pubSubProvider";
import { buildSchema } from "./schemaBuilder";
import { defaults } from "@gsox/core";

export function createServer(app, options) {
      const curOpt = {...defaults, ...options}
      const { host, port, routes, inject } = curOpt;
      const schema = buildSchema(inject);

      // add webhook endpoint
      app.post(routes.webhook, bodyParser.json(), (req, res) => {
            pubSub.publish(NOTIFICATION_TOPIC, req.body);
            return res.sendStatus(200);
      });

      // init apollo server
      const server = new ApolloServer({schema});
      server.applyMiddleware({ app });

      // setup websockets endpoint
      const ws = http(app);

      ws.listen({ host, port }, () => {
            // tslint:disable-next-line:no-console
            console.log(`ðŸš€ Server ready at http://${host}:${port}${server.graphqlPath}`);

      });
      SubscriptionServer.create({
            execute,
            schema,
            subscribe,
      }, {
            path: routes.graphql,
            server: ws,
      });
      return ws;
}
