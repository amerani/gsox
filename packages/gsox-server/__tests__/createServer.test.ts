import { Field, Type } from "@gsox/schema";
import * as express from "express";
import { applyMiddleware } from "../src/applyMiddleware";

@Type()
class Test {
      @Field("String")
      public success: boolean;
}

test("should create server", () => {
      expect(applyMiddleware(express(), { host: "localhost", port: 5000, inject: [Test] })).toMatchSnapshot();
});

test("should throw when not injected", () => {
      expect(() => applyMiddleware(express(), {})).toThrow();
});
