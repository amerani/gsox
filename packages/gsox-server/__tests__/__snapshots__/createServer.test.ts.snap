// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create server 1`] = `
Object {
  "apollo": ApolloServer {
    "context": undefined,
    "extensions": Array [],
    "graphqlPath": "/gsox/graphql",
    "playgroundOptions": Object {
      "settings": Object {
        "editor.fontFamily": "'Source Code Pro', 'Consolas', 'Inconsolata', 'Droid Sans Mono', 'Monaco', monospace",
        "editor.fontSize": 14,
        "editor.reuseHeaders": true,
        "editor.theme": "dark",
        "general.betaUpdates": false,
        "request.credentials": "omit",
        "tracing.hideTracingResponse": true,
      },
      "version": "1.7.2",
    },
    "requestOptions": Object {
      "cache": InMemoryLRUCache {
        "store": LRUCache {
          Symbol(max): Infinity,
          Symbol(lengthCalculator): [Function],
          Symbol(allowStale): false,
          Symbol(maxAge): 0,
          Symbol(dispose): undefined,
          Symbol(noDisposeOnSet): false,
          Symbol(cache): Map {},
          Symbol(lruList): Yallist {
            "head": null,
            "length": 0,
            "tail": null,
          },
          Symbol(length): 0,
        },
      },
      "persistedQueries": Object {
        "cache": InMemoryLRUCache {
          "store": LRUCache {
            Symbol(max): Infinity,
            Symbol(lengthCalculator): [Function],
            Symbol(allowStale): false,
            Symbol(maxAge): 0,
            Symbol(dispose): undefined,
            Symbol(noDisposeOnSet): false,
            Symbol(cache): Map {},
            Symbol(lruList): Yallist {
              "head": null,
              "length": 0,
              "tail": null,
            },
            Symbol(length): 0,
          },
        },
      },
    },
    "schema": GraphQLSchema {
      "__allowedLegacyNames": undefined,
      "_directives": Array [
        GraphQLDirective {
          "args": Array [
            Object {
              "astNode": undefined,
              "defaultValue": undefined,
              "description": "Skipped when true.",
              "name": "if",
              "type": "Boolean!",
            },
          ],
          "astNode": undefined,
          "description": "Directs the executor to skip this field or fragment when the \`if\` argument is true.",
          "locations": Array [
            "FIELD",
            "FRAGMENT_SPREAD",
            "INLINE_FRAGMENT",
          ],
          "name": "skip",
        },
        GraphQLDirective {
          "args": Array [
            Object {
              "astNode": undefined,
              "defaultValue": undefined,
              "description": "Included when true.",
              "name": "if",
              "type": "Boolean!",
            },
          ],
          "astNode": undefined,
          "description": "Directs the executor to include this field or fragment only when the \`if\` argument is true.",
          "locations": Array [
            "FIELD",
            "FRAGMENT_SPREAD",
            "INLINE_FRAGMENT",
          ],
          "name": "include",
        },
        GraphQLDirective {
          "args": Array [
            Object {
              "astNode": undefined,
              "defaultValue": "No longer supported",
              "description": "Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted in [Markdown](https://daringfireball.net/projects/markdown/).",
              "name": "reason",
              "type": "String",
            },
          ],
          "astNode": undefined,
          "description": "Marks an element of a GraphQL schema as no longer supported.",
          "locations": Array [
            "FIELD_DEFINITION",
            "ENUM_VALUE",
          ],
          "name": "deprecated",
        },
      ],
      "_implementations": Object {},
      "_mutationType": null,
      "_queryType": "Query",
      "_subscriptionType": "Subscription",
      "_typeMap": Object {
        "Boolean": "Boolean",
        "Int": "Int",
        "Ping": "Ping",
        "Query": "Query",
        "String": "String",
        "Subscription": "Subscription",
        "Test": "Test",
        "__Directive": "__Directive",
        "__DirectiveLocation": "__DirectiveLocation",
        "__EnumValue": "__EnumValue",
        "__Field": "__Field",
        "__InputValue": "__InputValue",
        "__Schema": "__Schema",
        "__Type": "__Type",
        "__TypeKind": "__TypeKind",
        "schema": "schema",
      },
      "astNode": undefined,
    },
    "subscriptionServerOptions": Object {
      "path": "/graphql",
    },
    "subscriptionsPath": "/graphql",
    "uploadsConfig": Object {},
  },
  "close": [Function],
  "host": "localhost",
  "listen": [Function],
  "port": 5000,
  "routes": Object {
    "graphql": "/gsox/graphql",
    "webhook": "/gsox/webhook",
  },
  "server": Server {
    "_connections": 0,
    "_events": Object {
      "connection": [Function],
      "error": [Function],
      "listening": [Function],
      "request": [Function],
      "upgrade": [Function],
    },
    "_eventsCount": 5,
    "_handle": null,
    "_maxListeners": undefined,
    "_pendingResponseData": 0,
    "_unref": false,
    "_usingWorkers": false,
    "_workers": Array [],
    "allowHalfOpen": true,
    "httpAllowHalfOpen": false,
    "keepAliveTimeout": 5000,
    "maxHeadersCount": null,
    "pauseOnConnect": false,
    "timeout": 120000,
    Symbol(IncomingMessage): [Function],
    Symbol(ServerResponse): [Function],
    Symbol(asyncId): -1,
  },
}
`;
